### ISSUE TITLE
QUESTION: Are Azure Functions supported with scaffold dapr task?

### ISSUE THREAD

**Issue by jsacapdev** [microsoft/vscode-dapr#97]
# issue

`Dapr: Scaffold Dapr Task` does not work with an Azure Function.

A git repository to replicate can be found at (its on the `main` branch, not `master`)

https://github.com/jsacapdev/fun.dapr.client

1. Clone the repository
2. cd into src/fundaprapi
3. `code .`
4. start debugging

# expected result

i can call an api and hit a breakpoint

# actual result

first time around this is the output:

```
> Executing task: daprd-debug <

> Executing command: daprd --app-id "app" --app-port "7071" --placement-address "localhost:50005" <

time="2020-06-09T08:46:18.56803982Z" level=info msg="starting Dapr Runtime -- version 0.7.0 -- commit 57bd1e6" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.56808572Z" level=info msg="log level set to: info" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.570090642Z" level=info msg="metrics server started on :9090/" app_id=app instance=techdev scope=dapr.metrics type=log ver=0.7.0
time="2020-06-09T08:46:18.572121664Z" level=info msg="loading default configuration" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.572189865Z" level=info msg="standalone mode configured" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.572205465Z" level=info msg="app id: app" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.572224265Z" level=info msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.573988184Z" level=info msg="found component statestore (state.redis)" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.573991484Z" level=info msg="found component messagebus (pubsub.redis)" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:46:18.574375588Z" level=info msg="application protocol: http. waiting on port 7071.  This will block until the app is listening on that port." app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
```

and then vscocde hangs:

```
Failed to detect running Functions host within "60" seconds. You may want to adjust the "azureFunctions.pickProcessTimeout" setting.
```

terminating the `daprd-debug` task and restarting then returns:

```
> Executing task: daprd-debug <

> Executing command: daprd --app-id "app" --app-port "7071" --placement-address "localhost:50005" <

time="2020-06-09T08:49:20.12068077Z" level=info msg="starting Dapr Runtime -- version 0.7.0 -- commit 57bd1e6" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.120717371Z" level=info msg="log level set to: info" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.120917572Z" level=info msg="metrics server started on :9090/" app_id=app instance=techdev scope=dapr.metrics type=log ver=0.7.0
time="2020-06-09T08:49:20.121309275Z" level=info msg="loading default configuration" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.121381376Z" level=info msg="standalone mode configured" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.121394176Z" level=info msg="app id: app" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.121409976Z" level=info msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.122041681Z" level=info msg="found component statestore (state.redis)" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.122132981Z" level=info msg="found component messagebus (pubsub.redis)" app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.122161682Z" level=info msg="application protocol: http. waiting on port 7071.  This will block until the app is listening on that port." app_id=app instance=techdev scope=dapr.runtime type=log ver=0.7.0
time="2020-06-09T08:49:20.121245675Z" level=fatal msg="failed to start metrics server: listen tcp :9090: bind: address already in use" app_id=app instance=techdev scope=dapr.metrics type=log ver=0.7.0

Terminal will be reused by tasks, press any key to close it.
```

# additional information

running this from the command line does work:

`dapr run --app-id fundaprapi --app-port 7071 --port 3500 func start`

vscode:

```
Version: 1.45.1
Commit: 5763d909d5f12fe19f215cbfdd29a91c0fa9208a
Date: 2020-05-14T08:27:22.494Z
Electron: 7.2.4
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Linux x64 5.3.0-1022-azure
```

**Comment by jsacapdev** [comment]
this is still a problem, with the release of the azure functions `Dapr.AzureFunctions.Extension`

**Comment by philliphoff** [comment]
@jsacapdev Thanks for the update. I've only just started to look into this as I've had time (which hasn't been much, lately) but run into issues with Azure Functions itself that I need to sort out.

**Comment by philliphoff** [comment]
@jsacapdev There are a couple of things going on here.  First, the Azure Functions project scaffolding does not start the functions host via a `preLaunchTask` in the debug launch configuration.  Instead, it implicitly runs the `func: host start` task as a result of the configuration using a Azure Functions command to pick the debug process.  Hence, the Dapr scaffolding doesn't know that it needs to chain in the `func: host start` task when it scaffolds the Dapr launch task.

We could add special logic that recognizes the standard Azure Functions scaffolding and do the appropriate chaining in that case.  However, this appears to lead to a fundamental VS Code issue ([vscode#70283](https://github.com/microsoft/vscode/issues/70283)) where a task cannot depend on another task that is a background task (which the `func: host start` task happens to be).  So, even when properly chained, VS Code will not recognize that the function host has started in order to allow the Dapr task to begin (which relies on being able to make requests to the function).

I'll need to think for a bit to see if there's some other way we could accommodate both the quirks of the Azure Functions scaffolding and the outstanding VS Code issue.
